services/mirror/mirror.go

```diff
@@ -78,21 +82,56 @@ func Update(ctx context.Context) error {
 			return nil
 		}
 
+		// Check we've not been cancelled
 		select {
 		case <-ctx.Done():
-			return fmt.Errorf("Aborted")
+			return fmt.Errorf("aborted")
 		default:
-			return mirrorQueue.Push(&item)
 		}
+
+		// Check if this request is already in the queue
+		has, err := mirrorQueue.Has(&item)
+		if err != nil {
+			return err
+		}
+		if has {
+			return nil
+		}
+
+		// Double check we've not been cancelled
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("aborted")
+		default:
+		}
+
+		// Push to the Queue
+		if err := mirrorQueue.Push(&item); err != nil {
+			return err
+		}
+
+		requested++
+		if limit > 0 && requested > limit {
+			return errLimit
+		}
+		return nil
 	}
 
-	if err := models.MirrorsIterate(handler); err != nil {
-		log.Error("MirrorsIterate: %v", err)
-		return err
+	if pullLimit != 0 {
+		if err := models.MirrorsIterate(func(idx int, bean interface{}) error {
+			return handler(idx, bean, pullLimit)
+		}); err != nil && err != errLimit {
+			log.Error("MirrorsIterate: %v", err)
+			return err
+		}
 	}
-	if err := models.PushMirrorsIterate(handler); err != nil {
-		log.Error("PushMirrorsIterate: %v", err)
-		return err
+	if pushLimit != 0 {
```

Comment:
Here and above: what does `limit == 0` mean? document only says that `negative values mean no limit`

Reply:
yes it will disable that particular type.

Reply:
Should we document this behavior or just leave it as it is?

Reply:
I don't think it is inconsistent with the current documentation

Reply:
But 0 is still very special here  (and some users may set PULL/PUSH limit to 0 by mistake).

For example, if a user set PUSH/PULL limit 0, then no error is reported, no pull/push  is executed, that's a big surprise. Even if they come to the discord channel to ask for help, we also have no idea at first glance.

So I think it's better to document this behavior, and make some clear log messages.